File: app/src/main/java/com/tk4218/grocerylistr/Model/GroceryListItem.java
package com.tk4218.grocerylistr.Model;

import android.util.Log;

import com.tk4218.grocerylistr.Database.JSONResult;
import com.tk4218.grocerylistr.Database.QueryBuilder;


public class GroceryListItem {
    private QueryBuilder mQb = new QueryBuilder();

    private int mGroceryListItemKey;
    private int mGroceryListKey;
    private Ingredient mIngredient;
    private MeasurementConverter mIngredientAmount = new MeasurementConverter();
    private boolean mAddedToCart;

    public GroceryListItem(int groceryListItemKey){
        JSONResult groceryListItem = mQb.getGroceryListItem(groceryListItemKey);
        if(groceryListItem.getCount() > 0){
            setGroceryListItemKey(groceryListItemKey);
            setGroceryListKey(groceryListItem.getInt("GroceryListKey"));
            setIngredient(groceryListItem.getInt("IngredientKey"));
            setIngredientUnit(groceryListItem.getString("IngredientUnit"));
            setIngredientAmount(groceryListItem.getDouble("IngredientAmount"));
            setAddedToCart(groceryListItem.getBoolean("AddedToCart"));
        }
    }

    public GroceryListItem(int groceryListItemKey, int groceryListKey, int ingredientKey, double ingredientAmount, String ingredientUnit, boolean addedToCart){
        setGroceryListItemKey(groceryListItemKey);
        setGroceryListKey(groceryListKey);
        setIngredient(ingredientKey);
        setIngredientUnit(ingredientUnit);
        setIngredientAmount(ingredientAmount);
        setAddedToCart(addedToCart);
    }

    public void setGroceryListItemKey(int groceryListItemKey){ mGroceryListItemKey = groceryListItemKey; }
    public int getGroceryListItemKey(){ return mGroceryListItemKey; }

    public void setGroceryListKey(int groceryListKey){ mGroceryListKey = groceryListKey; }
    public int getGroceryListKey(){ return mGroceryListKey; }

    public void setIngredient(int ingredientKey){ mIngredient = new Ingredient(ingredientKey); }
    public Ingredient getIngredient(){ return mIngredient; }

    public void setIngredientAmount(double amount){
        mIngredientAmount.setmMeasurementAmount(amount);
    }
    public boolean addIngredientAmount(double amount, String unit) {
        return mIngredientAmount.add(amount, unit);
    }
    public double getIngredientAmount(){
        return mIngredientAmount.getMeasurementAmount();
    }
    public String getFormattedIngredientAmount(){
        return toFraction(getIngredientAmount(), 10);
    }

    public void setIngredientUnit(String ingredientUnit){
        mIngredientAmount.setMesurementUnit(ingredientUnit);
    }
    public String getIngredientUnit(){
        return mIngredientAmount.getMeasurementUnit();
    }

    public void setAddedToCart(boolean addedToCart){ mAddedToCart = addedToCart;}
    public boolean getAddedToCart(){ return mAddedToCart; }

    private String toFraction(double d, int factor) {
        StringBuilder sb = new StringBuilder();
        if (d < 0) {
            sb.append('-');
            d = -d;
        }
        long l = (long) d;
        if (l != 0) sb.append(l);
        d -= l;
        double error = Math.abs(d);
        int bestDenominator = 1;
        for(int i=2;i<=factor;i++) {
            double error2 = Math.abs(d - (double) Math.round(d * i) / i);
            if (error2 < error) {
                error = error2;
                bestDenominator = i;
            }
        }
        if (bestDenominator > 1) {
            if (l != 0) sb.append(' ');
            sb.append(Math.round(d * bestDenominator)).append('/').append(bestDenominator);
        }
        return sb.toString();
    }

}

File: app/src/main/java/com/tk4218/grocerylistr/model/GroceryListItem.kt
package com.tk4218.grocerylistr.model

import com.google.firebase.database.FirebaseDatabase
import java.util.*

class GroceryListItem() {
    var username: String = ""
    var groceryListItemKey: String = ""
    var groceryListKey: String = ""
    var ingredient: Ingredient = Ingredient()
    private val mIngredientAmount = MeasurementConverter()
    var addedToCart: Boolean = false

    var ingredientAmount: Double
        get() = mIngredientAmount.measurementAmount
        set(amount) = mIngredientAmount.setmMeasurementAmount(amount)

    val formattedIngredientAmount: String
        get() = toFraction(ingredientAmount, 10)

    var ingredientUnit: String
        get() = mIngredientAmount.measurementUnit
        set(ingredientUnit) = mIngredientAmount.setMesurementUnit(ingredientUnit)

    constructor(username: String, groceryListKey: String, ingredient: Ingredient): this(){
        this.username = username
        this.groceryListKey = groceryListKey
        this.ingredient = ingredient
        this.ingredientAmount = ingredient.ingredientAmount
        this.ingredientUnit = ingredient.ingredientUnit
        groceryListItemKey = UUID.randomUUID().toString()
    }

    fun addIngredientAmount(amount: Double, unit: String): Boolean {
        return mIngredientAmount.add(amount, unit)
    }

    fun save() {
        val groceryListItemRef = FirebaseDatabase.getInstance().getReference("$username/groceryList/$groceryListKey/$groceryListItemKey")
        groceryListItemRef.setValue(this)
    }

    private fun toFraction(d: Double, factor: Int): String {
        var d = d
        val sb = StringBuilder()
        if (d < 0) {
            sb.append('-')
            d = -d
        }
        val l = d.toLong()
        if (l != 0L) sb.append(l)
        d -= l.toDouble()
        var error = Math.abs(d)
        var bestDenominator = 1
        for (i in 2..factor) {
            val error2 = Math.abs(d - Math.round(d * i).toDouble() / i)
            if (error2 < error) {
                error = error2
                bestDenominator = i
            }
        }
        if (bestDenominator > 1) {
            if (l != 0L) sb.append(' ')
            sb.append(Math.round(d * bestDenominator)).append('/').append(bestDenominator)
        }
        return sb.toString()
    }

}
